### PR TITLE
feat: `SquashDictInnerUsedAccessesAssert` hint

### LINKED ISSUE #301: SquashDictInnerUsedAccessesAssert
Asserts the number of used accesses for a key during squashing.

Reference: https://github.com/starkware-libs/cairo-lang/blob/efa9648f57568aad8f8a13fbf027d2de7c63c2c0/src/starkware/cairo/common/squash_dict.cairo#L206

Depends on #245

### PR DESCRIPTION
**Pull request by TAdev0**
Resolves #301 

This PR implements`SquashDictInnerUsedAccessesAssert` hint, which checks that `n_used_accesses` Cairo local variable is equal to the number of used accesses for a key during dict squashing.

### CODE REVIEW

**Comment by TAdev0** [comment]
also addressed error in the name of the hint in the description for `SquashDictInnerLenAssert`

File: pkg/hintrunner/zero/zerohint_dictionaries.go [hunk 1]
```diff
@@ -128,3 +129,52 @@ func createSquashDictInnerNextKeyHinter(resolver hintReferenceResolver) (hinter.
...
+	return &GenericZeroHinter{
+		Name: "SquashDictInnerUsedAccessesAssert",
+		Op: func(vm *VM.VirtualMachine, ctx *hinter.HintRunnerContext) error {
+			//> assert ids.n_used_accesses == len(access_indices[key])
+			access_indices_, err := ctx.ScopeManager.GetVariableValue("access_indices")
+			if err != nil {
+				return err
+			}
+
+			access_indices := access_indices_.(map[fp.Element][]fp.Element)
```
Comment (har777) [inline pkg/hintrunner/zero/zerohint_dictionaries.go:148]: Check for `!ok {}` here

File: pkg/hintrunner/zero/zerohint_dictionaries.go [hunk 2]
```diff
@@ -128,3 +129,52 @@ func createSquashDictInnerNextKeyHinter(resolver hintReferenceResolver) (hinter.
...
+
+			key_, err := ctx.ScopeManager.GetVariableValue("key")
+			if err != nil {
+				return err
+			}
+
+			key := key_.(fp.Element)
```
Comment (har777) [inline pkg/hintrunner/zero/zerohint_dictionaries.go:155]: Check for !ok {} here

File: pkg/hintrunner/zero/zerohint_dictionaries.go [hunk 3]
```diff
@@ -128,3 +129,52 @@ func createSquashDictInnerNextKeyHinter(resolver hintReferenceResolver) (hinter.
...
+
+			accessIndicesAtKey := uint64(len(access_indices[key]))
+
+			nUsedAccessesField, err := hinter.ResolveAsUint64(vm, nUsedAccesses)
+			if err != nil {
+				return err
+			}
+
+			if accessIndicesAtKey != nUsedAccessesField {
+				return fmt.Errorf("assertion ids.n_used_accesses == len(access_indices[key] failed")
```
Comment (har777) [inline pkg/hintrunner/zero/zerohint_dictionaries.go:165]: `len(access_indices[key]` is missing the closing bracket

Comment (TAdev0) [reply pkg/hintrunner/zero/zerohint_dictionaries.go:165]: true, thanks!

File: pkg/hintrunner/zero/zerohint_dictionaries.go (hunk 2 above)
Comment (TAdev0) [reply pkg/hintrunner/zero/zerohint_dictionaries.go:155]: thanks!

File: pkg/hintrunner/zero/zerohint_dictionaries.go (hunk 1 above)
Comment (TAdev0) [reply pkg/hintrunner/zero/zerohint_dictionaries.go:148]: thanks!

**Comment by TAdev0** [comment]
@har777  comments addressed!

File: pkg/hintrunner/zero/zerohint_dictionaries.go [hunk 4]
```diff
@@ -128,3 +129,58 @@ func createSquashDictInnerNextKeyHinter(resolver hintReferenceResolver) (hinter.
...
+	return &GenericZeroHinter{
+		Name: "SquashDictInnerUsedAccessesAssert",
+		Op: func(vm *VM.VirtualMachine, ctx *hinter.HintRunnerContext) error {
+			//> assert ids.n_used_accesses == len(access_indices[key])
+			access_indices_, err := ctx.ScopeManager.GetVariableValue("access_indices")
+			if err != nil {
+				return err
+			}
+
+			access_indices, ok := access_indices_.(map[fp.Element][]fp.Element)
```
Comment (har777) [inline pkg/hintrunner/zero/zerohint_dictionaries.go:148]: Looking at the squash dict functionality, I am pretty sure the array doesn't need to be an array of `[]fp.Element` and easily be something like `uint64` or even smaller.

This is how squash dict populates this object.
```
access_indices = {}
for i in range(n_accesses):
    key = memory[address + dict_access_size * i]
    access_indices.setdefault(key, []).append(i)
```

No way `i` is going to iterate from `0` to a super huge number. But I think for now this is fine. Just wanted to note this here for later.

**Review by har777 — APPROVED** [review]

File: pkg/hintrunner/zero/zerohint_dictionaries.go (hunk 4 above)
Comment (TAdev0) [reply pkg/hintrunner/zero/zerohint_dictionaries.go:148]: agree with you!

File: pkg/hintrunner/zero/zerohint_dictionaries.go [hunk 5]
```diff
@@ -128,3 +129,58 @@ func createSquashDictInnerNextKeyHinter(resolver hintReferenceResolver) (hinter.
...
+			if err != nil {
+				return err
+			}
+
+			key, ok := key_.(fp.Element)
+			if !ok {
+				return fmt.Errorf("cannot cast key_ to felt")
+			}
+
+			accessIndicesAtKey := uint64(len(access_indices[key]))
```
[the remark this answers is not in the document]
Comment (TAdev0) [reply pkg/hintrunner/zero/zerohint_dictionaries.go:163]: true!

**Comment by TAdev0** [comment]
@cicr99  @rodrigo-pino  ready to be merged

**Review by rodrigo-pino — APPROVED** [review]
LGTM

**Pull request merged**